source/agora/network/NetworkManager.d

```diff
@@ -835,8 +862,9 @@ public class NetworkManager
 
     ***************************************************************************/
 
-    private void getBlocksFrom (Height block_height,
-        scope bool delegate(const(Block)[]) @safe onReceivedBlocks) nothrow
+    private void getBlocksFrom (Ledger ledger, Height block_height,
+        scope bool delegate(Ledger ledger, const(Block)[], const(PreImageInfo)[]) @safe onReceivedBlocks)
+        nothrow
```

Comment:
I think this mixes things up quite a bit. How come we can receive a valid block but not know the pre-images ?

Reply:
> I think this mixes things up quite a bit. How come we can receive a valid block but not know the pre-images?

Before making a new block, we should know the pre-images as you know. So the following code will be running when the case you are saying happens which means getting blocks and pre-images from the next node. 

```
// one or more blocks were rejected, stop retrieval from node
if (!onReceivedBlocks(ledger, blocks, preimages))
    continue LNextNode;
```

I thought that there's a very low possibility that pre-images are not correct although blocks are valid.

In fact, there's one more thing we should think about it. What if there's no node from which we can get pre-images? We can imagine that there are nodes that only have the ledger and no normal database. Don't we need to store the last pre-images of un-enrolled validators into the ledger, including slashed validators? I know you don't want the ledger to be bloated due to that. I will file a new issue about that if needed.

Reply:
> I think this mixes things up quite a bit. How come we can receive a valid block but not know the pre-images?

Anyway, we need to mix blocks and pre-images up in this function because making a new block needs proper pre-images of the current height. The fact that a block received from some node is valid means the pre-images from the node also valid because the blocks were created with the right pre-images. If there is some invalid information, the catch-up process can restart getting valid information from other nodes at the point where the failure occurs. If you have any other suggestions about that, please let me know or we can make things right with another new issue.